---
title: "heritable"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{heritable}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE,
  message = FALSE,
  fig.width = 7, 
  fig.height = 5, 
  fig.align = "center",
  out.width = "80%"
  )
```

## Motivation

A key goal in any breeding trial is calculate heritability, which describes the extent of which a trait of interest underpinned by genes.

`heritable` is the one-stop shop for heritability calculations in R. Here, we have implemented existing methods for heritability to aid reproducibility and reporting of it's calculations. This vignette is a brief overview of `heritable`'s workflow and key features.

## Installation

Note that this package is under active development. You can install the development version of heritable from [GitHub](https://github.com/) with: 


```{r setup}
# install.packages("pak")
# pak::pak("anu-aagi/heritable")
library(heritable)
```

### An overview

`heritable` is compatible with model outputs from `asreml` and `lme4` and extracts the relevant variance components to calculate heritability. 

Let's work with  dataset that contains phenotypic measurements of downy mildew resistance of 89 lettuce genotypes across 3 locations (environments), with replicates. For demonstration purposes, we will use subset of single environment (`L2`) where genotypes are fully replicated. 

```{r}
library(dplyr)
head(lettuce_phenotypes)

lettuce_subset <- lettuce_phenotypes |> 
  filter(loc == "L2")
```

```{r, echo=FALSE}
library(ggplot2)

# Create the field layout plot
ggplot(lettuce_subset, aes(x = gen, y = rep)) +
  geom_tile(aes(fill = y), color = "white") +
  scale_fill_gradient(high = "#006d2c", low = "#d4eac7", 
                      name = "Downy Mildew\nResistance") +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, size = 5),
    legend.position = "bottom"
  )
```

We also have access to a genomic relationship matrix (GRM) calculated from 300 genetic markers that we will use for narrow-sense heritability. 

```{r}
# View the structure of the GRM
dim(lettuce_GRM)
lettuce_GRM[1:5, 1:5]
```

### Broad-sense heritability

Here we have provided code for both `ASREML-R` and `lme4` to fit a model with genotype as a random effect. **Note** that since this is a single environment and balanced subset of the data, all heritability estimates should be the same. 

```{r, eval=FALSE}
# Fit an asreml model with genotype as random effect
library(asreml)
lettuce_asreml <- asreml(
  fixed = y ~ rep,
  random = ~ gen,
  data = lettuce_subset, 
  trace = FALSE
)

# Fit an lme4 model with genotype as random effect
library(lme4)
lettuce_lme4 <- lmer(y ~ rep + (1 | gen), data = lettuce_subset)
```


```{r, eval=FALSE}
# Calculate broad-sense heritability using multiple methods
H2(lettuce_asreml, target = "gen", method = c("Standard", "Cullis", "Oakey"))
H2(lettuce_lme4, target = "gen", method = c("Standard", "Cullis", "Oakey"))
```

### Narrow-sense heritability with asreml and GRM

Narrow-sense heritability is currently only implemented for `asreml`. 
In the model, we will use the `lettuce_GRM` genomic relationship matrix and fit it using `asreml::vm()`:

```{r, eval=FALSE}
# Fit model with GRM for narrow-sense heritability
lettuce_asreml_grm <- asreml(
  fixed = y ~ loc,
  random = ~ vm(gen, lettuce_GRM) + rep,
  data = lettuce_phenotypes, 
  trace = FALSE
)

# Calculate narrow-sense heritability
h2(lettuce_asreml_grm, target = "gen", method = "Oakey")
```

## Heritability methods

`heritable` supports calculations of both narrow-sense ($h^2$) and broad-sense heritability ($H^2$) using a variety of methods for single environment breeding trials:
